src/objective/rank_obj.cu

```diff
@@ -0,0 +1,651 @@
+    dh::LaunchN(device_id_, niter, nullptr, [=] __device__(size_t idx) {
+      // First, determine the group 'idx' belongs to
+      unsigned item_idx = idx % total_items;
+      int group_idx = dh::UpperBound(dgroups, ngroups, item_idx);
+      // Span of this group within the larger labels/predictions sorted tuple
+      int group_begin = dgroups[group_idx - 1];
+      int group_end = dgroups[group_idx];
+      int total_group_items = group_end - group_begin;
+
+      // Are the labels diverse enough? If they are all the same, then there is nothing to pick
+      // from another group - bail sooner
+      if (sorted_labels[group_begin] == sorted_labels[group_end - 1]) return;
+
+      // Find the number of labels less than and greater than the current label
+      // at the sorted index position item_idx
+      int nleft  = LowerBound(sorted_labels + group_begin, total_group_items,
+                              sorted_labels[item_idx]);
+      int nright = UpperBound(sorted_labels + group_begin, total_group_items,
+                              sorted_labels[item_idx]);
+      nleft  = (nleft  == -1) ? 0 : nleft + 1;
+      nright = (nright == -1) ? 0 : total_group_items - nright;
+
+      // Create a minstd_rand object to act as our source of randomness
+      thrust::minstd_rand rng;
+      rng.discard(idx);
+      // Create a uniform_int_distribution to produce a sample from outside of the
+      // present label group
+      thrust::uniform_int_distribution<int> dist(0, nleft + nright - 1);
+
+      int sample = dist(rng);
+      int pos_idx = -1;  // Bigger label
+      int neg_idx = -1;  // Smaller label
+      // Are we picking a sample to the left/right of the current group?
+      if (sample < nleft) {
+        // Go left
+        pos_idx = sample + group_begin;
+        neg_idx = item_idx;
+      } else {
+        pos_idx = item_idx;
+        int items_in_group = total_group_items - nleft - nright;
+        neg_idx = sample + items_in_group + group_begin;
+      }
+
+      // Compute and assign the gradients now
+      const float eps = 1e-16f;
+      bst_float p = common::Sigmoid(dpreds[original_pos[pos_idx]] - dpreds[original_pos[neg_idx]]);
+      bst_float g = p - 1.0f;
+      bst_float h = thrust::max(p * (1.0f - p), eps);
+
+      // Rescale each gradient and hessian so that the group has a weighted constant
+      float scale = 1.0f / (niter / total_items);
+      if (fix_list_weight != 0.0f) {
+        scale *= fix_list_weight / total_group_items;
+      }
+
+      float weight = num_weights ? dweights[group_idx - 1] : 1.0f;
+      weight *= weight_normalization_factor;
+      weight *= scale;
+      // Accumulate gradient and hessian in both positive and negative indices
+      float *out_pos_gpair = reinterpret_cast<float *>(&out_gpair[original_pos[pos_idx]]);
+      const GradientPair in_pos_gpair(g * weight, 2.0f * weight * h);
+      atomicAdd(out_pos_gpair, in_pos_gpair.GetGrad());
```

Comment:
There's an `AtomicAddGpair` in `updater_gpu_common.cuh`, Can we reuse that?

Reply:
done